### ISSUE TITLE
Visual Studio debugger does not sometimes recognize local variables in rewritten assembly

### ISSUE THREAD

**Issue by manison** [Microsoft/CodeContracts#434]
Debugger in Visual Studio 2015 sometimes does not recognize local variables in assemblies with applied Code Contracts. In such case the debugger's _Locals_ windows does not list local variables at all, hovering over the variable produces no tooltip and putting the variable name into the _Watch_ window outputs _The name does not exist in the current context_. When the _Perform Runtime Contract Checking_ check box is unchecked in the affected project then the debugger shows variables as expected.

![ccnok](https://cloud.githubusercontent.com/assets/2073329/15642902/834dbfd0-264a-11e6-8d87-b3f62d2a8c36.PNG)

![ccok](https://cloud.githubusercontent.com/assets/2073329/15642916/94de33e2-264a-11e6-92b6-7f21c239b213.PNG)

Code Contracts version used: 1.9.10714.2

**Comment by krk** [comment]
+1

**Comment by SergeyTeplyakov** [comment]
Is there any chances to get a reproducable solution? And is it the case that the same codebase with no ccrewrite is working fine?

**Comment by manison** [comment]
> And is it the case that the same codebase with no ccrewrite is working fine?

Correct. When the same project is rebuilt with ccrewrite not being involved (_Perform Runtime Contract Checking_ check box is unchecked) then the debugging experience is fine.

I tried to isolate a reproducible solution however I could not come up with anything yet. The (static) method that exhibited the wrong debugger behavior now works fine whereas another (instance) method now shows the odd behavior. Async methods are involved in the call chain. Any ideas what to focus on?

**Comment by SergeyTeplyakov** [comment]
Just to brainstorm: is there any preconditions/postconditions in async methods?

P.S. Without small repo, it would be really hard to solve the problem:(

**Comment by manison** [comment]
> Just to brainstorm: is there any preconditions/postconditions in async methods?

There are plenty of Requires.

> P.S. Without small repo, it would be really hard to solve the problem:(

That's for sure. I'm just asking for any hint since the codebase is quite large and the issue seems rather random for now.

**Comment by SergeyTeplyakov** [comment]
Ok, you can think of a mix of features:

Do you have async and iterators in the same file? Maybe it is a struct, not a class? Maybe there are some object invariants as well? Is there inheritance involved? Preconditions for the current types constructor that has a base type?

All those are complicated features inside and mix of them could lead to some issues.

**Comment by Daniel-Svensson** [comment]
@manison I have just uploaded a small simple program to compare pdb information for two assemblies (https://github.com/Daniel-Svensson/PdbComparer). Maybe you can extend it to also compare symbol informations for variables?

The program is provided as is, I've tried it to compare a few assemblies but never against an actual assembly with problems (We fixed them). Hope it can help

**Comment by manison** [comment]
@Daniel-Svensson I modified your application to also compare local variables inside methods (https://github.com/manison/PdbComparer). I'm not 100% sure if the comparison algorithm I implemented is correct since I'm not familiar with Mono.Cecil. However with this method I wasn't able to find any difference between the problematic rewritten assembly and the original one.

In the meantime I found somewhat similar issue https://connect.microsoft.com/VisualStudio/feedback/details/1026319/local-variables-in-an-async-method-arent-showing-values-via-tooltips. Can there be some relation?

**Comment by johncrim** [comment]
+1

Are you still having difficulty reproducing this? My team and I see this problem all the time, it's a real hassle to workaround when debugging.

**Comment by manison** [comment]
I still hit the issue from time to time. However I haven't been able to find correlation between rewritten and not rewritten assembly yet. Have you tried the PdbComparer utility?
